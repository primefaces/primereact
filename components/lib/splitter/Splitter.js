import React, { forwardRef, memo, useRef } from 'react';
import PropTypes from 'prop-types';
import { DomHandler, classNames, ObjectUtils } from '../utils/Utils';
import { useMountEffect, useEventListener } from '../hooks/Hooks';

export const SplitterPanel = () => { }

export const Splitter = memo(forwardRef((props, ref) => {
    const elementRef = useRef(null);
    const gutterRef = useRef(null);
    const size = useRef(null);
    const dragging = useRef(null);
    const startPos = useRef(null);
    const prevPanelElement = useRef(null);
    const nextPanelElement = useRef(null);
    const prevPanelSize = useRef(null);
    const nextPanelSize = useRef(null);
    const prevPanelIndex = useRef(null);
    const panelSizes = useRef(null);
    const isStateful = props.stateKey != null;

    const [bindDocumentMouseMoveListener, unbindDocumentMouseMoveListener] = useEventListener({ type: 'mousemove', listener: (event) => onResize(event) });
    const [bindDocumentMouseUpListener, unbindDocumentMouseUpListener] = useEventListener({
        type: 'mouseup', listener: event => {
            onResizeEnd(event);
            unbindMouseListeners();
        }
    });

    const bindMouseListeners = () => {
        bindDocumentMouseMoveListener();
        bindDocumentMouseUpListener();
    }

    const unbindMouseListeners = () => {
        unbindDocumentMouseMoveListener();
        unbindDocumentMouseUpListener();
    }

    const validateResize = (newPrevPanelSize, newNextPanelSize) => {
        if (props.children[prevPanelIndex.current].props && props.children[prevPanelIndex.current].props.minSize && props.children[prevPanelIndex.current].props.minSize > newPrevPanelSize) {
            return false;
        }

        if (props.children[prevPanelIndex.current + 1].props && props.children[prevPanelIndex.current + 1].props.minSize && props.children[prevPanelIndex.current + 1].props.minSize > newNextPanelSize) {
            return false;
        }

        return true;
    }

    const clear = () => {
        dragging.current = false;
        size.current = null;
        startPos.current = null;
        prevPanelElement.current = null;
        nextPanelElement.current = null;
        prevPanelSize.current = null;
        nextPanelSize.current = null;
        prevPanelIndex.current = null;
    }

    const getStorage = () => {
        switch (props.stateStorage) {
            case 'local':
                return window.localStorage;

            case 'session':
                return window.sessionStorage;

            default:
                throw new Error(props.stateStorage + ' is not a valid value for the state storage, supported values are "local" and "session".');
        }
    }

    const saveState = () => {
        getStorage().setItem(props.stateKey, JSON.stringify(panelSizes));
    }

    const restoreState = () => {
        const storage = getStorage();
        const stateString = storage.getItem(props.stateKey);

        if (stateString) {
            panelSizes.current = JSON.parse(stateString);
            let children = [...elementRef.current.children].filter(child => DomHandler.hasClass(child, 'p-splitter-panel'));
            children.forEach((child, i) => {
                child.style.flexBasis = 'calc(' + panelSizes[i] + '% - ' + ((props.children.length - 1) * props.gutterSize) + 'px)';
            });

            return true;
        }

        return false;
    }

    const onResizeStart = (event, index) => {
        let pageX = event.type === 'touchstart' ? event.touches[0].pageX : event.pageX;
        let pageY = event.type === 'touchstart' ? event.touches[0].pageY : event.pageY;
        size.current = props.layout === 'horizontal' ? DomHandler.getWidth(elementRef.current) : DomHandler.getHeight(elementRef.current);
        dragging.current = true;
        startPos.current = props.layout === 'horizontal' ? pageX : pageY;
        prevPanelElement.current = gutterRef.current.previousElementSibling;
        nextPanelElement.current = gutterRef.current.nextElementSibling;
        prevPanelSize.current = 100 * (props.layout === 'horizontal' ? DomHandler.getOuterWidth(prevPanelElement.current, true) : DomHandler.getOuterHeight(prevPanelElement.current, true)) / size.current;
        nextPanelSize.current = 100 * (props.layout === 'horizontal' ? DomHandler.getOuterWidth(nextPanelElement.current, true) : DomHandler.getOuterHeight(nextPanelElement.current, true)) / size.current;
        prevPanelIndex.current = index;
        DomHandler.addClass(gutterRef.current, 'p-splitter-gutter-resizing');
        DomHandler.addClass(elementRef.current, 'p-splitter-resizing');
    }

    const onResize = (event) => {
        let newPos;
        let pageX = event.type === 'touchmove' ? event.touches[0].pageX : event.pageX;
        let pageY = event.type === 'touchmove' ? event.touches[0].pageY : event.pageY;

        if (props.layout === 'horizontal')
            newPos = (pageX * 100 / size.current) - (startPos.current * 100 / size.current);
        else
            newPos = (pageY * 100 / size.current) - (startPos.current * 100 / size.current);

        let newPrevPanelSize = prevPanelSize.current + newPos;
        let newNextPanelSize = nextPanelSize.current - newPos;

        if (validateResize(newPrevPanelSize, newNextPanelSize)) {
            prevPanelElement.current.style.flexBasis = 'calc(' + newPrevPanelSize + '% - ' + ((props.children.length - 1) * props.gutterSize) + 'px)';
            nextPanelElement.current.style.flexBasis = 'calc(' + newNextPanelSize + '% - ' + ((props.children.length - 1) * props.gutterSize) + 'px)';
            panelSizes.current[prevPanelIndex.current] = newPrevPanelSize;
            panelSizes.current[prevPanelIndex.current + 1] = newNextPanelSize;
        }
    }

    const onResizeEnd = (event) => {
        if (isStateful) {
            saveState();
        }

        if (props.onResizeEnd) {
            props.onResizeEnd({
                originalEvent: event,
                sizes: panelSizes
            })
        }

        DomHandler.removeClass(gutterRef.current, 'p-splitter-gutter-resizing');
        DomHandler.removeClass(elementRef.current, 'p-splitter-resizing');
        clear();
    }

    const onGutterMouseDown = (event, index) => {
        onResizeStart(event, index);
        bindMouseListeners();
    }

    const onGutterTouchStart = (event, index) => {
        onResizeStart(event, index);

        window.addEventListener('touchmove', onGutterTouchMove, { passive: false, cancelable: false });
        window.addEventListener('touchend', onGutterTouchEnd);
    }

    const onGutterTouchMove = (event) => {
        onResize(event);
    }

    const onGutterTouchEnd = (event) => {
        onResizeEnd(event);
        window.removeEventListener('touchmove', onGutterTouchMove);
        window.removeEventListener('touchend', onGutterTouchEnd);
    }

    useMountEffect(() => {
        let panelElements = [...elementRef.current.children].filter(child => DomHandler.hasClass(child, 'p-splitter-panel'));
        panelElements.map(panelElement => {
            if (panelElement.childNodes && ObjectUtils.isNotEmpty(DomHandler.find(panelElement, '.p-splitter'))) {
                DomHandler.addClass(panelElement, 'p-splitter-panel-nested');
            }
        });

        if (props.children && props.children.length) {
            let initialized = false;
            if (isStateful) {
                initialized = restoreState();
            }

            if (!initialized) {
                let _panelSizes = [];

                props.children.map((panel, i) => {
                    let panelInitialSize = panel.props && panel.props.size ? panel.props.size : null;
                    let panelSize = panelInitialSize || (100 / props.children.length);
                    _panelSizes[i] = panelSize;
                    panelElements[i].style.flexBasis = 'calc(' + panelSize + '% - ' + ((props.children.length - 1) * props.gutterSize) + 'px)';
                    return _panelSizes;
                });

                panelSizes.current = _panelSizes;
            }
        }
    });

    const createPanel = (panel, index) => {
        const panelClassName = classNames('p-splitter-panel', panel.props.className);
        const gutterStyle = props.layout === 'horizontal' ? { width: props.gutterSize + 'px' } : { height: props.gutterSize + 'px' }
        const gutter = (index !== props.children.length - 1) && (
            <div ref={gutterRef} className="p-splitter-gutter" style={gutterStyle} onMouseDown={event => onGutterMouseDown(event, index)}
                onTouchStart={event => onGutterTouchStart(event, index)} onTouchMove={event => onGutterTouchMove(event)} onTouchEnd={event => onGutterTouchEnd(event)}>
                <div className="p-splitter-gutter-handle"></div>
            </div>
        );

        return (
            <React.Fragment>
                <div key={index} className={panelClassName} style={panel.props.style}>
                    {panel.props.children}
                </div>
                {gutter}
            </React.Fragment>
        )
    }


    const createPanels = () => {
        return React.Children.map(props.children, createPanel);
    }

    const className = classNames(`p-splitter p-component p-splitter-${props.layout}`, props.className);
    const panels = createPanels();

    return (
        <div ref={elementRef} id={props.id} className={className} style={props.style}>
            {panels}
        </div>
    )
}));

SplitterPanel.defaultProps = {
    __TYPE: 'SplitterPanel',
    size: null,
    minSize: null,
    style: null,
    className: null
}

SplitterPanel.propTypes /* remove-proptypes */ = {
    __TYPE: PropTypes.string,
    header: PropTypes.number,
    minSize: PropTypes.number,
    style: PropTypes.object,
    className: PropTypes.string
}

Splitter.defaultProps = {
    __TYPE: 'Splitter',
    id: null,
    className: null,
    style: null,
    layout: 'horizontal',
    gutterSize: 4,
    stateKey: null,
    stateStorage: 'session',
    onResizeEnd: null
}

Splitter.propTypes /* remove-proptypes */ = {
    __TYPE: PropTypes.string,
    id: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.object,
    layout: PropTypes.string,
    gutterSize: PropTypes.number,
    stateKey: PropTypes.string,
    stateStorage: PropTypes.string,
    onResizeEnd: PropTypes.func
}
