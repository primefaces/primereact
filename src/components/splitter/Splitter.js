import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { classNames } from '../utils/ClassNames';
import DomHandler from '../utils/DomHandler';

export class SplitterPanel extends Component {

    static defaultProps = {
        size: null,
        minSize: null,
        style: null,
        className: null
    }

    static propTypes = {
        header: PropTypes.number,
        minSize: PropTypes.number,
        style: PropTypes.object,
        className: PropTypes.string
    }
}

export class Splitter extends Component {

    static defaultProps = {
        id: null,
        className: null,
        style: null,
        layout: 'horizontal',
        gutterSize: 4,
        stateKey: null,
        stateStorage: 'session',
        onResizeEnd: null
    }

    static propTypes = {
        id: PropTypes.string,
        className: PropTypes.string,
        style: PropTypes.object,
        layout: PropTypes.string,
        gutterSize: PropTypes.number,
        stateKey: PropTypes.string,
        stateStorage: PropTypes.string,
        onResizeEnd: PropTypes.func
    };

    bindMouseListeners() {
        if (!this.mouseMoveListener) {
            this.mouseMoveListener = event => this.onResize(event)
            document.addEventListener('mousemove', this.mouseMoveListener);
        }

        if (!this.mouseUpListener) {
            this.mouseUpListener = event => {
                this.onResizeEnd(event);
                this.unbindMouseListeners();
            }
            document.addEventListener('mouseup', this.mouseUpListener);
        }
    }

    validateResize(newPrevPanelSize, newNextPanelSize) {
        if (this.props.children[0].props && this.props.children[0].props.minSize && this.props.children[0].props.minSize > newPrevPanelSize) {
            return false;
        }

        if (this.props.children[1].props && this.props.children[1].props.minSize && this.props.children[1].props.minSize > newNextPanelSize) {
            return false;
        }

        return true;
    }

    unbindMouseListeners() {
        if (this.mouseMoveListener) {
            document.removeEventListener('mousemove', this.mouseMoveListener);
            this.mouseMoveListener = null;
        }

        if (this.mouseUpListener) {
            document.removeEventListener('mouseup', this.mouseUpListener);
            this.mouseUpListener = null;
        }
    }

    clear() {
        this.dragging = false;
        this.size = null;
        this.startPos = null;
        this.prevPanelElement = null;
        this.nextPanelElement = null;
        this.prevPanelSize = null;
        this.nextPanelSize = null;
        this.gutterElement = null;
        this.prevPanelIndex = null;
    }

    isStateful() {
        return this.props.stateKey != null;
    }

    getStorage() {
        switch (this.props.stateStorage) {
            case 'local':
                return window.localStorage;

            case 'session':
                return window.sessionStorage;

            default:
                throw new Error(this.props.stateStorage + ' is not a valid value for the state storage, supported values are "local" and "session".');
        }
    }

    saveState() {
        this.getStorage().setItem(this.props.stateKey, JSON.stringify(this.panelSizes));
    }

    restoreState() {
        const storage = this.getStorage();
        const stateString = storage.getItem(this.props.stateKey);

        if (stateString) {
            this.panelSizes = JSON.parse(stateString);
            let children = [...this.container.children].filter(child => DomHandler.hasClass(child, 'p-splitter-panel'));
            children.forEach((child, i) => {
                child.style.flexBasis = 'calc(' + this.panelSizes[i] + '% - ' + ((this.props.children.length - 1) * this.props.gutterSize) + 'px)';
            });

            return true;
        }

        return false;
    }

    onResizeStart(event, index) {
        this.gutterElement = event.currentTarget;
        this.size = this.horizontal ? DomHandler.getWidth(this.container) : DomHandler.getHeight(this.container);
        this.dragging = true;
        this.startPos = this.props.layout === 'horizontal' ? event.pageX : event.pageY;
        this.prevPanelElement = this.gutterElement.previousElementSibling;
        this.nextPanelElement = this.gutterElement.nextElementSibling;
        this.prevPanelSize = 100 * (this.props.layout === 'horizontal' ? DomHandler.getOuterWidth(this.prevPanelElement, true) : DomHandler.getOuterHeight(this.prevPanelElement, true)) / this.size;
        this.nextPanelSize = 100 * (this.props.layout === 'horizontal' ? DomHandler.getOuterWidth(this.nextPanelElement, true) : DomHandler.getOuterHeight(this.nextPanelElement, true)) / this.size;
        this.prevPanelIndex = index;
        DomHandler.addClass(this.gutterElement, 'p-splitter-gutter-resizing');
        DomHandler.addClass(this.container, 'p-splitter-resizing');
    }

    onResize(event) {
        let newPos;
        if (this.props.layout === 'horizontal')
            newPos = (event.pageX * 100 / this.size) - (this.startPos * 100 / this.size);
        else
            newPos = (event.pageY * 100 / this.size) - (this.startPos * 100 / this.size);

        let newPrevPanelSize = this.prevPanelSize + newPos;
        let newNextPanelSize = this.nextPanelSize - newPos;

        if (this.validateResize(newPrevPanelSize, newNextPanelSize)) {
            this.prevPanelElement.style.flexBasis = 'calc(' + newPrevPanelSize + '% - ' + ((this.props.children.length - 1) * this.props.gutterSize) + 'px)';
            this.nextPanelElement.style.flexBasis = 'calc(' + newNextPanelSize + '% - ' + ((this.props.children.length - 1) * this.props.gutterSize) + 'px)';
            this.panelSizes[this.prevPanelIndex] = newPrevPanelSize;
            this.panelSizes[this.prevPanelIndex + 1] = newNextPanelSize;
        }
    }

    onResizeEnd(event) {
        if (this.isStateful()) {
            this.saveState();
        }

        if (this.props.onResizeEnd) {
            this.props.onResizeEnd({
                originalEvent: event,
                sizes: this.panelSizes
            })
        }

        DomHandler.removeClass(this.gutterElement, 'p-splitter-gutter-resizing');
        DomHandler.removeClass(this.container, 'p-splitter-resizing');
        this.clear();
    }

    onGutterMouseDown(event, index) {
        this.onResizeStart(event, index);
        this.bindMouseListeners();
    }

    onGutterTouchStart(event, index) {
        this.onResizeStart(event, index);

        this.windowTouchMoveListener = this.onGutterTouchMove.bind(this);
        this.windowTouchEndListener = this.onGutterTouchEnd.bind(this);
        window.addEventListener('touchmove', this.windowTouchMoveListener, { passive: false, cancelable: false });
        window.addEventListener('touchend', this.windowTouchEndListener);

    }

    onGutterTouchMove(event) {
        this.onResize(event);
    }

    onGutterTouchEnd(event) {
        this.onResizeEnd(event);

        window.removeEventListener('touchmove', this.windowTouchMoveListener);
        window.removeEventListener('touchend', this.windowTouchEndListener);
        this.windowTouchMoveListener = null;
        this.windowTouchEndListener = null;
    }

    componentDidMount() {
        if (this.panelElement) {
            if (this.panelElement.childNodes && DomHandler.find(this.panelElement, '.p-splitter')) {
                DomHandler.addClass(this.panelElement, 'p-splitter-panel-nested');
            }
        }

        if (this.props.children && this.props.children.length) {
            let initialized = false;
            if (this.isStateful()) {
                initialized = this.restoreState();
            }

            if (!initialized) {
                let children = [...this.container.children].filter(child => DomHandler.hasClass(child, 'p-splitter-panel'));
                let _panelSizes = [];

                this.props.children.map((panel, i) => {
                    let panelInitialSize = panel.props && panel.props.size ? panel.props.size : null;
                    let panelSize = panelInitialSize || (100 / this.props.children.length);
                    _panelSizes[i] = panelSize;
                    children[i].style.flexBasis = 'calc(' + panelSize + '% - ' + ((this.props.children.length - 1) * this.props.gutterSize) + 'px)';
                    return _panelSizes;
                });

                this.panelSizes = _panelSizes;
            }
        }
    }

    renderPanel(panel, index) {
        const panelClassName = classNames('p-splitter-panel', panel.props.className);
        const gutterStyle = this.props.layout === 'horizontal' ? { width: this.props.gutterSize + 'px' } : { height: this.props.gutterSize + 'px' }
        const gutter = (index !== this.props.children.length - 1) && (
            <div ref={(el) => this.gutterElement = el} className={'p-splitter-gutter'} style={gutterStyle} onMouseDown={event => this.onGutterMouseDown(event, index)}
                onTouchStart={event => this.onGutterTouchStart(event, index)} onTouchMove={event => this.onGutterTouchMove(event)} onTouchEnd={event => this.onGutterTouchEnd(event)}>
                <div className="p-splitter-gutter-handle"></div>
            </div>
        );

        return (
            <React.Fragment>
                <div ref={(el) => this.panelElement = el} key={index} className={panelClassName} style={panel.props.style}>
                    {panel.props.children}
                </div>
                {gutter}
            </React.Fragment>
        );
    }

    renderPanels() {
        return (
            React.Children.map(this.props.children, (panel, index) => {
                return this.renderPanel(panel, index);
            })
        )
    }

    render() {
        const className = classNames(`p-splitter p-component p-splitter-${this.props.layout}`, this.props.className);
        const panels = this.renderPanels();

        return (
            <div ref={(el) => this.container = el} id={this.props.id} className={className} style={this.props.style}>
                {panels}
            </div>
        );
    }
}
