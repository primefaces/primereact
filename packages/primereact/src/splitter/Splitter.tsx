'use client';
import { Component, withComponent } from '@primereact/core/component';
import { useSplitter } from '@primereact/headless/splitter';
import { styles } from '@primereact/styles/splitter';
import { mergeProps } from '@primeuix/utils';
import * as React from 'react';
import { defaultProps } from './Splitter.props';
import { SplitterGutter } from './gutter';
import { SplitterPanel } from './panel';

export const Splitter = withComponent({
    defaultProps,
    styles,
    setup: (instance) => {
        const splitter = useSplitter(instance.inProps);

        return splitter;
    },
    render: (instance) => {
        const {
            id,
            props,
            ptmi,
            cx,
            // element refs
            elementRef
        } = instance;

        const rootProps = mergeProps(
            {
                id,
                className: cx('root'),
                'data-p-resizing': false
            },
            ptmi('root')
        );

        return (
            <Component as={props.as || 'div'} {...rootProps} ref={elementRef}>
                {props.children}
            </Component>
        );
    },
    components: {
        Panel: SplitterPanel,
        Gutter: SplitterGutter
    }
});
