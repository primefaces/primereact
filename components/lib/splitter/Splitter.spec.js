import '@testing-library/jest-dom';
import { snapshot } from '../../test';
import { Splitter, SplitterPanel } from './Splitter';

describe('Splitter', () => {
    snapshot(
        <>
            <Splitter style={{ height: '300px' }} className="mb-5">
                <SplitterPanel className="flex align-items-center justify-content-center">Panel 1</SplitterPanel>
                <SplitterPanel className="flex align-items-center justify-content-center">Panel 2</SplitterPanel>
            </Splitter>
        </>,
        'Splitter requires two SplitterPanel components to wrap.'
    );
    snapshot(
        <>
            <Splitter style={{ height: '300px' }} layout="vertical">
                <SplitterPanel>Panel 1</SplitterPanel>
                <SplitterPanel>Panel 2</SplitterPanel>
            </Splitter>
        </>,
        'Vertical layout'
    );
    snapshot(
        <>
            <Splitter style={{ height: '300px' }}>
                <SplitterPanel className="flex align-items-center justify-content-center" size={20} minSize={10}>
                    Panel 1
                </SplitterPanel>
                <SplitterPanel size={80}>
                    <Splitter layout="vertical">
                        <SplitterPanel className="flex align-items-center justify-content-center" size={15}>
                            Panel 2
                        </SplitterPanel>
                        <SplitterPanel size={85}>
                            <Splitter>
                                <SplitterPanel className="flex align-items-center justify-content-center" size={20}>
                                    Panel 3
                                </SplitterPanel>
                                <SplitterPanel className="flex align-items-center justify-content-center" size={80}>
                                    Panel 4
                                </SplitterPanel>
                            </Splitter>
                        </SplitterPanel>
                    </Splitter>
                </SplitterPanel>
            </Splitter>
        </>,
        'Nested'
    );
    snapshot(
        <>
            <Splitter style={{ height: '300px' }} layout="vertical">
                <SplitterPanel size={5}>Panel 1</SplitterPanel>
            </Splitter>
        </>,
        'Single Panel with size'
    );
    snapshot(
        <>
            <Splitter style={{ height: '300px' }} layout="vertical">
                <SplitterPanel>Panel 1</SplitterPanel>
            </Splitter>
        </>,
        'Single Panel without size'
    );
    snapshot(
        <>
            <Splitter style={{ height: '300px' }} layout="vertical"></Splitter>
        </>,
        'Without panels'
    );
});
