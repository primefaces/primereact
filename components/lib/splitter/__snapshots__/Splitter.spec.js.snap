// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splitter Nested 1`] = `
<div>
  <div
    class="p-splitter p-component p-splitter-horizontal"
    style="height: 300px;"
  >
    <div
      class="p-splitter-panel flex align-items-center justify-content-center"
      style="flex-basis: calc(20% - 4px);"
    >
      Panel 1
    </div>
    <div
      class="p-splitter-gutter"
      style="width: 4px;"
    >
      <div
        class="p-splitter-gutter-handle"
      />
    </div>
    <div
      class="p-splitter-panel p-splitter-panel-nested"
      style="flex-basis: calc(80% - 4px);"
    >
      <div
        class="p-splitter p-component p-splitter-vertical"
      >
        <div
          class="p-splitter-panel flex align-items-center justify-content-center"
          style="flex-basis: calc(15% - 4px);"
        >
          Panel 2
        </div>
        <div
          class="p-splitter-gutter"
          style="height: 4px;"
        >
          <div
            class="p-splitter-gutter-handle"
          />
        </div>
        <div
          class="p-splitter-panel p-splitter-panel-nested"
          style="flex-basis: calc(85% - 4px);"
        >
          <div
            class="p-splitter p-component p-splitter-horizontal"
          >
            <div
              class="p-splitter-panel flex align-items-center justify-content-center"
              style="flex-basis: calc(20% - 4px);"
            >
              Panel 3
            </div>
            <div
              class="p-splitter-gutter"
              style="width: 4px;"
            >
              <div
                class="p-splitter-gutter-handle"
              />
            </div>
            <div
              class="p-splitter-panel flex align-items-center justify-content-center"
              style="flex-basis: calc(80% - 4px);"
            >
              Panel 4
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter Single Panel with size 1`] = `
<div>
  <div
    class="p-splitter p-component p-splitter-vertical"
    style="height: 300px;"
  >
    <div
      class="p-splitter-panel"
      style="flex-basis: calc(5% - 0px);"
    >
      Panel 1
    </div>
    <div
      class="p-splitter-gutter"
      style="height: 4px;"
    >
      <div
        class="p-splitter-gutter-handle"
      />
    </div>
  </div>
</div>
`;

exports[`Splitter Single Panel without size 1`] = `
<div>
  <div
    class="p-splitter p-component p-splitter-vertical"
    style="height: 300px;"
  >
    <div
      class="p-splitter-panel"
      style="flex-basis: calc(100% - 0px);"
    >
      Panel 1
    </div>
    <div
      class="p-splitter-gutter"
      style="height: 4px;"
    >
      <div
        class="p-splitter-gutter-handle"
      />
    </div>
  </div>
</div>
`;

exports[`Splitter Splitter requires two SplitterPanel components to wrap. 1`] = `
<div>
  <div
    class="p-splitter p-component p-splitter-horizontal mb-5"
    style="height: 300px;"
  >
    <div
      class="p-splitter-panel flex align-items-center justify-content-center"
      style="flex-basis: calc(50% - 4px);"
    >
      Panel 1
    </div>
    <div
      class="p-splitter-gutter"
      style="width: 4px;"
    >
      <div
        class="p-splitter-gutter-handle"
      />
    </div>
    <div
      class="p-splitter-panel flex align-items-center justify-content-center"
      style="flex-basis: calc(50% - 4px);"
    >
      Panel 2
    </div>
  </div>
</div>
`;

exports[`Splitter Vertical layout 1`] = `
<div>
  <div
    class="p-splitter p-component p-splitter-vertical"
    style="height: 300px;"
  >
    <div
      class="p-splitter-panel"
      style="flex-basis: calc(50% - 4px);"
    >
      Panel 1
    </div>
    <div
      class="p-splitter-gutter"
      style="height: 4px;"
    >
      <div
        class="p-splitter-gutter-handle"
      />
    </div>
    <div
      class="p-splitter-panel"
      style="flex-basis: calc(50% - 4px);"
    >
      Panel 2
    </div>
  </div>
</div>
`;

exports[`Splitter Without panels 1`] = `
<div>
  <div
    class="p-splitter p-component p-splitter-vertical"
    style="height: 300px;"
  />
</div>
`;
